src/protocol/msgcodec.c

```diff
@@ -307,108 +295,91 @@ int8_t _z_res_decl_encode(_z_wbuf_t *wbf, uint8_t header, const _z_res_decl_t *d
     int8_t ret = _Z_RES_OK;
     _Z_DEBUG("Encoding _Z_DECL_RESOURCE\n");
 
-    // Encode the body
     _Z_EC(_z_zint_encode(wbf, dcl->_id))
-    ret = _z_keyexpr_encode(wbf, header, &dcl->_key);
+    ret |= _z_keyexpr_encode(wbf, header, &dcl->_key);
 
     return ret;
 }
 
-void _z_res_decl_decode_na(_z_zbuf_t *zbf, uint8_t header, _z_res_decl_result_t *r) {
+int8_t _z_res_decl_decode_na(_z_res_decl_t *dcl, _z_zbuf_t *zbf, uint8_t header) {
     _Z_DEBUG("Decoding _Z_DECL_RESOURCE\n");
-    r->_tag = _Z_RES_OK;
+    int8_t ret = _Z_RES_OK;
 
-    _z_zint_result_t r_zint = _z_zint_decode(zbf);
-    _ASSURE_P_RESULT(r_zint, r, _Z_ERR_PARSE_ZINT)
-    r->_value._id = r_zint._value;
+    ret |= _z_zint_decode(&dcl->_id, zbf);
+    ret |= _z_keyexpr_decode(&dcl->_key, zbf, header);
 
-    _z_keyexpr_result_t r_res = _z_keyexpr_decode(zbf, header);
-    _ASSURE_P_RESULT(r_res, r, _Z_ERR_PARSE_RESKEY)
-    r->_value._key = r_res._value;
+    return ret;
 }
 
-_z_res_decl_result_t _z_res_decl_decode(_z_zbuf_t *zbf, uint8_t header) {
-    _z_res_decl_result_t r;
-    _z_res_decl_decode_na(zbf, header, &r);
-    return r;
+int8_t _z_res_decl_decode(_z_res_decl_t *dcl, _z_zbuf_t *zbf, uint8_t header) {
+    return _z_res_decl_decode_na(dcl, zbf, header);
 }
 
 /*------------------ Publisher Declaration ------------------*/
 int8_t _z_pub_decl_encode(_z_wbuf_t *wbf, uint8_t header, const _z_pub_decl_t *dcl) {
     int8_t ret = _Z_RES_OK;
     _Z_DEBUG("Encoding _Z_DECL_PUBLISHER\n");
 
-    // Encode the body
-    ret = _z_keyexpr_encode(wbf, header, &dcl->_key);
+    ret |= _z_keyexpr_encode(wbf, header, &dcl->_key);
 
     return ret;
 }
 
-void _z_pub_decl_decode_na(_z_zbuf_t *zbf, uint8_t header, _z_pub_decl_result_t *r) {
+int8_t _z_pub_decl_decode_na(_z_pub_decl_t *dcl, _z_zbuf_t *zbf, uint8_t header) {
     _Z_DEBUG("Decoding _Z_DECL_PUBLISHER\n");
-    r->_tag = _Z_RES_OK;
+    int8_t ret = _Z_RES_OK;
+
+    ret |= _z_keyexpr_decode(&dcl->_key, zbf, header);
 
-    _z_keyexpr_result_t r_res = _z_keyexpr_decode(zbf, header);
-    _ASSURE_P_RESULT(r_res, r, _Z_ERR_PARSE_RESKEY)
-    r->_value._key = r_res._value;
+    return ret;
 }
 
-_z_pub_decl_result_t _z_pub_decl_decode(_z_zbuf_t *zbf, uint8_t header) {
-    _z_pub_decl_result_t r;
-    _z_pub_decl_decode_na(zbf, header, &r);
-    return r;
+int8_t _z_pub_decl_decode(_z_pub_decl_t *dcl, _z_zbuf_t *zbf, uint8_t header) {
+    return _z_pub_decl_decode_na(dcl, zbf, header);
 }
 
 /*------------------ Subscriber Declaration ------------------*/
 int8_t _z_sub_decl_encode(_z_wbuf_t *wbf, uint8_t header, const _z_sub_decl_t *dcl) {
     int8_t ret = _Z_RES_OK;
     _Z_DEBUG("Encoding _Z_DECL_SUBSCRIBER\n");
 
-    // Encode the body
     _Z_EC(_z_keyexpr_encode(wbf, header, &dcl->_key))
     if (_Z_HAS_FLAG(header, _Z_FLAG_Z_S) == true) {
-        ret = _z_subinfo_encode(wbf, &dcl->_subinfo);
+        ret |= _z_subinfo_encode(wbf, &dcl->_subinfo);
```

Comment:
If an error occurs on the decoding, it must always overlap the success (that happen before or after). Also, it allow us to detect if any system error also occurred, which should be treated differently from other Zenoh errors. 

Before we were taking the suggested approach, but together with early returns in case of errors. With MISRA, we need to go all the way to the single return in the end of the function. I found this approach to be cleaner in terms of code, but also got a better performance than with nested `IFs`.